File: frontend/src/lib/events.ts
// Mirrors backend/api/events.py - the LLM <-> Graph contract.

export interface TokenEvent { type: "token"; text: string; }
export interface ToolEvent { type: "tool"; name: string; status: "start" | "done" | "error"; detail: string; }
export interface GraphNodeEvent {
  type: "graph_node"; id: string; node_type: string; label: string; metrics: Record<string, any>;
}
export interface GraphEdgeEvent {
  type: "graph_edge"; source: string; target: string; edge_type: string; weight: number;
}
export interface NodeUpdateEvent { type: "node_update"; id: string; metrics: Record<string, any>; }
export interface RecommendationEvent { type: "recommendation"; data: Recommendation; }
export interface ErrorEvent { type: "error"; message: string; }
export interface DoneEvent { type: "done"; mode: string; }

export type TpoEvent =
  | TokenEvent | ToolEvent | GraphNodeEvent | GraphEdgeEvent
  | NodeUpdateEvent | RecommendationEvent | ErrorEvent | DoneEvent;

export interface Recommendation {
  product_id: number;
  label: string;
  recommended_depth: number;
  recommended_depth_pct: number;
  net_incremental_margin: number;
  clearance_value: number;
  total_value: number;
  net_roi: number;
  blended_margin_pct: number;
  margin_ok: boolean;
  inventory: number;
  weeks_of_supply: number;
  rationale: string;
  impact: {
    own_lift_units: number;
    cannibalization_margin: number;
    halo_margin: number;
    net_incremental_margin: number;
    top_cannibalized: { product_id: number; label: string; lost_units: number; lost_margin: number }[];
    top_halo: { product_id: number; label: string; gained_units: number; gained_margin: number }[];
  };
}

File: backend/api/events.py
"""Typed SSE event schema - the LLM <-> Graph contract.

Every entity the assistant references is emitted as one of these typed events,
so the frontend renders the graph from structured data (never by parsing prose).
Mirrored in frontend/src/lib/events.ts.
"""
from __future__ import annotations

from typing import Literal

from pydantic import BaseModel


class TokenEvent(BaseModel):
    type: Literal["token"] = "token"
    text: str


class ToolEvent(BaseModel):
    type: Literal["tool"] = "tool"
    name: str
    status: Literal["start", "done", "error"]
    detail: str = ""


class GraphNodeEvent(BaseModel):
    type: Literal["graph_node"] = "graph_node"
    id: str
    node_type: str
    label: str
    metrics: dict = {}


class GraphEdgeEvent(BaseModel):
    type: Literal["graph_edge"] = "graph_edge"
    source: str
    target: str
    edge_type: str
    weight: float = 1.0


class NodeUpdateEvent(BaseModel):
    """Update metrics on an existing node (e.g. colour by cannibalization)."""
    type: Literal["node_update"] = "node_update"
    id: str
    metrics: dict = {}


class RecommendationEvent(BaseModel):
    type: Literal["recommendation"] = "recommendation"
    data: dict


class ErrorEvent(BaseModel):
    type: Literal["error"] = "error"
    message: str


class DoneEvent(BaseModel):
    type: Literal["done"] = "done"
    mode: str = ""  # "llm" | "fallback"


Event = (
    TokenEvent | ToolEvent | GraphNodeEvent | GraphEdgeEvent
    | NodeUpdateEvent | RecommendationEvent | ErrorEvent | DoneEvent
)


def to_sse(event: BaseModel) -> dict:
    """Shape for sse_starlette.EventSourceResponse."""
    return {"event": event.type, "data": event.model_dump_json()}
